### DISCUSSION TITLE
Problem with transit-costs for TSP

### DISCUSSION THREAD

**Discussion by hsu-ret** [google/or-tools#3524] [Routing (and legacy CP) questions]
Hello everyone, 
I dont know, if I got something wrong about the TSP examples or if it's just not working correctly. I'll try to explain, what I did: 

**What version of OR-Tools and what language are you using?**
Version: main
Language: Python, running via PowerShell on Windows

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
Routing-Solver, 

**What did you do?**
I'm currently trying to set up a TSP for a 3D environment. For that i used the example

[https://developers.google.com/optimization/routing/tsp#circuit_board](url)

and changed the input points to an array of ~400 (3x1) tuples - to represent 3D points - and the calculation for euclidean-distance to 
```py
1000*math.hypot((int(from_node[0]) - int(to_node[0])),
                               (int(from_node[1]) - int(to_node[1])),
                               (int(from_node[2]) - int(to_node[2]))
                               )))
```
I  basically added the 3rd coordinate and a factor of 1000, so that the distance matrix would be more accurate in terms of diagonals.
Since this gave me a solution with many unneccessary ups and downs, i tried adding a penalty for the 3rd component with: 
```py
if int(from_node[2])!= int(to_node[2]):
    distances[from_counter][to_counter] += pz
```

where `pZ` can be like `pZ=10.000` to eliminate as many of the up and down movements as possible

**What did you expect to see**
I expected to see a change in the solution of the TSP. 

**What did you see instead?**
Whatever changes I make to the calculation of the `distance_matrix` - even setting all distances to 1 instead of the presented `hypot`-fct - results in a different `distance_matrix`, but does not change the solution of the TSP in any way. 

So now I'm wondering, if I'm getting anything wrong here, since i thought, the costs for the transit between 2 nodes are calculated based on the `distance_matrix`.? 
Or do I need to change anything in the `search_parameters` for this approach to work? I tried this with both the example options ("FirstSolution" & "LocalSearchMetaheuristics")

Thanks in advance, I'd appretiate any help.

**Comment by Mizux** [comment]
1) does your matrix distance are integer ?
2) can you provide a gist/repo with your examples and data ?
3) pZ should be `10_000` ? at least one or two order of magnitude higher than your distance between point so solver has real incentive to avoid it. aka changing plan is 10x/100x time costly than the distance between any two points....

**Reply by hsu-ret** [reply]
Hey, thanks for your fast reply. 

I never did this before, but I hope it works: [url](
https://gist.github.com/hsu-ret/c79ca03a6b0e644df03b5f97c6c26a52)

This is the code, I am using on the python side. The `return_script.py` is just some kind of a workaround, since I need the solution from the TSP in a matlab-environment. It takes the input from matlab ( [pyInput.zip](https://github.com/google/or-tools/files/9889523/pyInput.zip) ), generates the desired tuples-form of the TSP-example and passes them to the tsp.py main. 

As you can see, the pZ was just a spelling mistake by me, the penalty is 10 times  higher (1e3 vs. 1e4) than the smallest distance in my example.

**Comment by lperron** [comment]
can you check the output of your distance callback ? It might not be what you expect, or what the solve expects.

**Reply by hsu-ret** [reply]
Hey, yes the output of the distance_callback is correct. Both the penalty and the scaling factor are correctly applied.

**Comment by hsu-ret** [comment]
EDIT: I recently did some changes on the matlab-side and some tests on the number of points/nodes taken into account.  I have no idea, what exactly changed, because my .py-scripts still look exactly the same, but for now I get a reasonable solution. (the path is avoiding the movement in z-axis, as expected because of the penalty `pZ`)
I guess this post wont be helpful for anyone in the future, but thank you guys for your time and help.
